src/rsa/padding.rs

```diff
@@ -154,3 +155,150 @@ impl Verification for PKCS1 {
+            let digest_len = self.digest_alg.output_len;
+
+            // Step 2.
+            let m_hash = digest::digest(self.digest_alg,
+                                        msg.as_slice_less_safe());
+
+            // Step 3: where we assume the digest and salt are of equal length.
+            if em_len < 2 + (2 * digest_len) {
+                return Err(error::Unspecified)
+            }
+
+            // Steps 4 and 5: Parse encoded message as:
+            //     masked_db || h_hash || 0xbc
+            let db_len = em_len - digest_len - 1;
+            let masked_db = try!(em.skip_and_get_input(db_len));
+            let h_hash = try!(em.skip_and_get_input(digest_len));
+            if try!(em.read_byte()) != 0xbc {
+                return Err(error::Unspecified);
+            }
+
+            // Step 7.
+            let mut db = [0u8; MAX_OUTPUT_LEN];
+            let db = &mut db[..db_len];
+
+            try!(mgf1(self.digest_alg, h_hash.as_slice_less_safe(), db));
+
+            try!(masked_db.read_all(error::Unspecified, |masked_bytes| {
+                // Step 6. Check the top bits of first byte are zero.
+                let b = try!(masked_bytes.read_byte());
+                if b & !top_byte_mask != 0 {
+                    return Err(error::Unspecified);
+                }
+                db[0] ^= b;
+
+                // Step 8.
+                for i in 1..db.len() {
+                    db[i] ^= try!(masked_bytes.read_byte());
+                }
+                Ok(())
+            }));
+
+            // Step 9.
+            db[0] &= top_byte_mask;
+
+            // Step 10.
+            let pad_len = db.len() - digest_len - 1;
+            for i in 0..pad_len {
+                if db[i] != 0 {
+                    return Err(error::Unspecified);
+                }
+            }
+            if db[pad_len] != 1 {
+                return Err(error::Unspecified);
+            }
+
+            // Step 11.
+            let salt = &db[db.len() - digest_len..];
+
+            // Step 12 and 13: compute hash value of:
+            //     (0x)00 00 00 00 00 00 00 00 || m_hash || salt
+            let mut ctx = digest::Context::new(self.digest_alg);
+            ctx.update(&PSS_PREFIX_ZEROS);
+            ctx.update(m_hash.as_ref());
+            ctx.update(salt);
+            let h_hash_check = ctx.finish();
+
+            // Step 14.
+            if h_hash != h_hash_check.as_ref() {
+                return Err(error::Unspecified);
+            }
+            Ok(())
+        })
+    }
+}
+
+// Mask-generating function MGF1 as described in
+// https://tools.ietf.org/html/rfc3447#appendix-B.2.1.
+fn mgf1(digest_alg: &'static digest::Algorithm, seed: &[u8], mask: &mut [u8])
+        -> Result<(), error::Unspecified> {
+    let digest_len = digest_alg.output_len;
+
+    // Maximum counter value is the value of (mask_len / digest_len) rounded up.
+    let ctr_max = (mask.len() - 1) / digest_len;
```

Comment:
We should get rid of this code duplication between the loop and the block that immediately follows the loop. Maybe we can do this instead?:

````
for (i, mask_chunk) in mask.chunks_mut(digest_len).enumerate() {
    let mut ctx = digest::Context::new(digest_alg);
    ctx.update(seed);
    ctx.update(&polyfill::slice::be_u8_from_u32(i as u32));
    let digest = ctx.finish();
    mask_chunk.copy_from_slice(digest.as_ref()[..mask_chunk.len()];
}